ccl/changefeedccl: TestChangefeedSchemaChangeAllowBackfill failed

ccl/changefeedccl.TestChangefeedSchemaChangeAllowBackfill [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4672429&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4672429&tab=artifacts#/) on master @ [d2a588523715f914906f08b1365f8fc9c8ae2805](https://github.com/cockroachdb/cockroach/commits/d2a588523715f914906f08b1365f8fc9c8ae2805):


```
=== RUN   TestChangefeedSchemaChangeAllowBackfill
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/a77002d7c9453d7cd2d382f907780e13/logTestChangefeedSchemaChangeAllowBackfill1739784690
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestChangefeedSchemaChangeAllowBackfill
    changefeed_test.go:1611: -- test log scope end --
--- FAIL: TestChangefeedSchemaChangeAllowBackfill (37.46s)
=== RUN   TestChangefeedSchemaChangeAllowBackfill/kafka/add_column_computed
    changefeed_test.go:1487: 
        	Error Trace:	helpers_test.go:159
        	            				helpers_test.go:211
        	            				changefeed_test.go:1487
        	Error:      	Received unexpected error:
        	            	lease-insert: duplicate key value violates unique constraint "primary"
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).FetchTerminalJobErr
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:352
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).jobFailed
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:268
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*reportErrorResumer).OnFailOrCancel
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:218
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func3
        	            	  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1218
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
        	            	  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1219
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
        	            	  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1168
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob
        	            	  | 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:408
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:942
        	            	  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (2) lease-insert: duplicate key value violates unique constraint "primary"
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestChangefeedSchemaChangeAllowBackfill/kafka/add_column_computed
        --- FAIL: TestChangefeedSchemaChangeAllowBackfill/kafka/add_column_computed (0.18s)
=== RUN   TestChangefeedSchemaChangeAllowBackfill/kafka
    --- FAIL: TestChangefeedSchemaChangeAllowBackfill/kafka (4.45s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedSchemaChangeAllowBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-14147

Epic CRDB-11732

--- a/pkg/sql/catalog/lease/BUILD.bazel
+++ b/pkg/sql/catalog/lease/BUILD.bazel
@@ -31,6 +31,8 @@ go_library(
         "//pkg/sql/catalog/descpb",
         "//pkg/sql/catalog/internal/catkv",
         "//pkg/sql/catalog/nstree",
+        "//pkg/sql/pgwire/pgcode",
+        "//pkg/sql/pgwire/pgerror",
         "//pkg/sql/sem/tree",
         "//pkg/sql/sessiondata",
         "//pkg/sql/sqlutil",

--- a/pkg/sql/catalog/lease/storage.go
+++ b/pkg/sql/catalog/lease/storage.go
@@ -26,6 +26,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
 	"github.com/cockroachdb/cockroach/pkg/util/grpcutil"
@@ -167,11 +169,15 @@ func (s storage) acquire(
 	for r := retry.StartWithCtx(ctx, retry.Options{}); r.Next(); {
 		err := s.db.Txn(ctx, acquireInTxn)
 		var pErr *roachpb.AmbiguousResultError
-		if errors.As(err, &pErr) {
+		switch {
+		case errors.As(err, &pErr):
 			log.Infof(ctx, "ambiguous error occurred during lease acquisition for %v, retrying: %v", id, err)
 			continue
-		}
-		if err != nil {
+		case pgerror.GetPGCode(err) == pgcode.UniqueViolation:
+			log.Infof(ctx, "uniqueness violation occurred due to concurrent lease"+
+				" removal for %v, retrying: %v", id, err)
+			continue
+		case err != nil:
 			return nil, hlc.Timestamp{}, err
 		}
 		log.VEventf(ctx, 2, "storage acquired lease %v@%v", desc, expiration)

